cmd/govim/gopls_client.go

```diff
@@ -225,11 +226,57 @@ func (g *govimplugin) PublishDiagnostics(ctxt context.Context, params *protocol.
 
 func (g *govimplugin) Progress(ctxt context.Context, params *protocol.ProgressParams) error {
 	defer absorbShutdownErr()
-	panic("Progress not implemented yet")
+	g.logGoplsClientf("Progress callback: %v", pretty.Sprint(params))
+
+	var ok bool
+	var raw map[string]interface{}
+	if raw, ok = params.Value.(map[string]interface{}); !ok {
+		return fmt.Errorf("unexpected value type: %T", params.Value)
+	}
+	var kind string
+	if kind, ok = raw["kind"].(string); !ok { // required by LSP
+		return fmt.Errorf("expected required field 'kind' in progress callback")
+	}
+	message, _ := raw["message"].(string) // optional
+	message = strings.TrimSpace(message)
+
+	var title string
+	if title, ok = raw["title"].(string); !ok && kind == "begin" { // required for "begin"
+		return fmt.Errorf("expected required field 'title' not set")
+	}
+
+	g.Schedule(func(govim.Govim) error {
+		v := g.vimstate
+
+		if v.config.ExperimentalProgressPopups == nil || !*v.config.ExperimentalProgressPopups {
```

Comment:
Per comment above, we don't actually need this check if we only enable progress as required.

Otherwise, if we can't drive this based on the capability, then I suggest we check the config value before scheduling (using the config lock - see elsewhere). It's relatively expensive to schedule only to do nothing.

Reply:
Done, now checking config value before schedule.